The `asm!` macro is not allowed in naked functions now

After [this PR](https://github.com/rust-lang/rust/pull/128651) merged in nightly toolchain, `asm!` macro is not allowed in naked functions. This affects code in chapter 5.

```
~/code/Asynchronous-Programming-in-Rust/ch05/c-fibers > cargo +nightly run                              
   Compiling c-fibers v0.1.0 (/home/yct21/code/Asynchronous-Programming-in-Rust/ch05/c-fibers)
error[E0787]: the `asm!` macro is not allowed in naked functions
   --> src/main.rs:144:5
    |
144 |     asm!("ret", options(noreturn))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ consider using the `naked_asm!` macro instead

error[E0787]: the `asm!` macro is not allowed in naked functions
   --> src/main.rs:158:5
    |
158 | /     asm!(
159 | |         "mov [rdi + 0x00], rsp",
160 | |         "mov [rdi + 0x08], r15",
161 | |         "mov [rdi + 0x10], r14",
...   |
174 | |         options(noreturn)
175 | |     );
    | |_____^ consider using the `naked_asm!` macro instead
```

--- a/ch05/c-fibers/src/main.rs
+++ b/ch05/c-fibers/src/main.rs
@@ -1,5 +1,13 @@
+/// FIX #31:
+/// Inline assembly blocks inside naked functions now need to use
+/// the `naked_asm` macro instead of the good old `asm` macro.
+/// The `noreturn` option is implicitly set by the `naked_asm`
+/// macro so there is no need to set that.
+///
+/// See: https://github.com/PacktPublishing/Asynchronous-Programming-in-Rust/issues/31
+/// for more information.
 #![feature(naked_functions)]
-use std::arch::asm;
+use std::arch::{asm, naked_asm};
 
 const DEFAULT_STACK_SIZE: usize = 1024 * 1024 * 2;
 const MAX_THREADS: usize = 4;
@@ -141,7 +149,7 @@ fn guard() {
 
 #[naked]
 unsafe extern "C" fn skip() {
-    asm!("ret", options(noreturn))
+    naked_asm!("ret")
 }
 
 pub fn yield_thread() {
@@ -155,7 +163,7 @@ pub fn yield_thread() {
 #[no_mangle]
 #[cfg_attr(target_os = "macos", export_name = "\x01switch")] // see: How-to-MacOS-M.md for explanation
 unsafe extern "C" fn switch() {
-    asm!(
+    naked_asm!(
         "mov [rdi + 0x00], rsp",
         "mov [rdi + 0x08], r15",
         "mov [rdi + 0x10], r14",
@@ -170,8 +178,7 @@ unsafe extern "C" fn switch() {
         "mov r12, [rsi + 0x20]",
         "mov rbx, [rsi + 0x28]",
         "mov rbp, [rsi + 0x30]",
-        "ret",
-        options(noreturn)
+        "ret"
     );
 }
 

--- a/ch05/d-fibers-closure/src/main.rs
+++ b/ch05/d-fibers-closure/src/main.rs
@@ -1,5 +1,13 @@
+/// FIX #31:
+/// Inline assembly blocks inside naked functions now need to use
+/// the `naked_asm` macro instead of the good old `asm` macro.
+/// The `noreturn` option is implicitly set by the `naked_asm`
+/// macro so there is no need to set that.
+///
+/// See: https://github.com/PacktPublishing/Asynchronous-Programming-in-Rust/issues/31
+/// for more information.
 #![feature(naked_functions)]
-use std::{arch::asm, ptr};
+use std::{arch::{asm, naked_asm}};
 
 const DEFAULT_STACK_SIZE: usize = 1024 * 1024 * 2;
 const MAX_THREADS: usize = 4;
@@ -152,8 +160,8 @@ fn call(thread: u64) {
 }
 
 #[naked]
-unsafe fn skip() {
-    asm!("ret", options(noreturn))
+unsafe extern "C" fn skip() {
+    naked_asm!("ret")
 }
 
 // this function is changed
@@ -175,8 +183,8 @@ pub fn yield_thread() {
 #[naked]
 #[no_mangle]
 #[cfg_attr(target_os = "macos", export_name = "\x01switch")]
-unsafe fn switch() {
-    asm!(
+unsafe extern "C" fn switch() {
+    naked_asm!(
         "mov 0x00[rdi], rsp",
         "mov 0x08[rdi], r15",
         "mov 0x10[rdi], r14",
@@ -192,8 +200,7 @@ unsafe fn switch() {
         "mov rbx, 0x28[rsi]",
         "mov rbp, 0x30[rsi]",
         "mov rdi, 0x38[rsi]",
-        "ret",
-        options(noreturn)
+        "ret"
     );
 }
 

--- a/ch05/e-fibers-windows/README.md
+++ b/ch05/e-fibers-windows/README.md
@@ -248,7 +248,7 @@ Last we need to change our `switch()`function and update our assembly. After all
 #[naked]
 #[no_mangle]
 unsafe extern "C" fn switch() {
-    asm!(
+    naked_asm!(
         "movaps      [rcx + 0x00], xmm6",
         "movaps      [rcx + 0x10], xmm7",
         "movaps      [rcx + 0x20], xmm8",
@@ -295,7 +295,7 @@ unsafe extern "C" fn switch() {
         "mov         gs:0x08, rax",
         "mov         rax, [rdx + 0xf0]",
         "mov         gs:0x10, rax",
-        "ret", options(noreturn)
+        "ret"
     );
 }
 ```

--- a/ch05/e-fibers-windows/src/main.rs
+++ b/ch05/e-fibers-windows/src/main.rs
@@ -1,5 +1,13 @@
+/// FIX #31:
+/// Inline assembly blocks inside naked functions now need to use
+/// the `naked_asm` macro instead of the good old `asm` macro.
+/// The `noreturn` option is implicitly set by the `naked_asm`
+/// macro so there is no need to set that.
+///
+/// See: https://github.com/PacktPublishing/Asynchronous-Programming-in-Rust/issues/31
+/// for more information.
 #![feature(naked_functions)]
-use std::arch::asm;
+use std::arch::{asm, naked_asm};
 
 const DEFAULT_STACK_SIZE: usize = 1024 * 1024 * 2;
 const MAX_THREADS: usize = 4;
@@ -145,7 +153,7 @@ impl Runtime {
 
 #[naked]
 unsafe extern "C" fn skip() {
-    asm!("ret", options(noreturn))
+    naked_asm!("ret")
 }
 
 
@@ -168,7 +176,7 @@ pub fn yield_thread() {
 #[no_mangle]
 #[cfg_attr(target_os = "macos", export_name = "\x01switch")]
 unsafe extern "C" fn switch() {
-    asm!(
+    naked_asm!(
         "mov [rdi + 0x00], rsp",
         "mov [rdi + 0x08], r15",
         "mov [rdi + 0x10], r14",
@@ -183,7 +191,7 @@ unsafe extern "C" fn switch() {
         "mov r12, [rsi + 0x20]",
         "mov rbx, [rsi + 0x28]",
         "mov rbp, [rsi + 0x30]",
-        "ret", options(noreturn)
+        "ret"
     );
 }
 

